libs/core/synchronization/include/hpx/synchronization/range_mutex.hpp

```diff
@@ -25,89 +25,114 @@ namespace hpx::synchronization {
     template <typename RangeMutex>
     class range_guard
     {
-        std::reference_wrapper<RangeMutex> lockRef;
-        std::size_t lockId = 0;
+        std::reference_wrapper<RangeMutex> mutex_ref;
+        std::size_t lock_id = 0;
 
     public:
         range_guard(RangeMutex& lock, std::size_t begin, std::size_t end)
-          : lockRef(lock)
+          : mutex_ref(lock)
         {
-            lockId = lockRef.get().lock(begin, end);
+            lock_id = mutex_ref.get().lock(begin, end);
         }
         ~range_guard()
         {
-            lockRef.get().unlock(lockId);
+            mutex_ref.get().unlock(lock_id);
+        }
+
+        range_guard(range_guard<RangeMutex>&) = delete;
+        range_guard<RangeMutex>& operator=(range_guard<RangeMutex>&) = delete;
+
+        range_guard(range_guard<RangeMutex>&& rhs_lock)
+        {
+            mutex_ref.get().unlock(lock_id);
+            mutex_ref = rhs_lock.mutex_ref;
+            lock_id = rhs_lock.lock_id;
+            rhs_lock.lock_id = 0;
+        }
+
+        range_guard<RangeMutex>& operator=(range_guard<RangeMutex>&& rhs_lock)
+        {
+            mutex_ref.get().unlock(lock_id);
+            mutex_ref = rhs_lock.mutex_ref;
+            lock_id = rhs_lock.lock_id;
+            rhs_lock.lock_id = 0;    // invalidating rhs_lock
         }
     };
 
     template <typename RangeMutex>
     class range_unique_lock
     {
-        std::reference_wrapper<RangeMutex> lockRef;
-        std::size_t lockId = 0;
+        std::reference_wrapper<RangeMutex> mutex_ref;
+        std::size_t lock_id = 0;
 
     public:
         range_unique_lock(RangeMutex& lock, std::size_t begin, std::size_t end)
-          : lockRef(lock)
+          : mutex_ref(lock)
         {
-            lockId = lockRef.get().lock(begin, end);
+            lock_id = mutex_ref.get().lock(begin, end);
         }
 
         ~range_unique_lock()
         {
-            lockRef.get().unlock(lockId);
+            mutex_ref.get().unlock(lock_id);
         }
 
-        void operator=(range_unique_lock<RangeMutex>&& lock)
+        range_unique_lock(range_guard<RangeMutex>&) = delete;
+        range_unique_lock<RangeMutex>& operator=(
+            range_unique_lock<RangeMutex>) = delete;
+
+        range_unique_lock<RangeMutex>& operator=(
+            range_unique_lock<RangeMutex>&& rhs_lock)
         {
-            lockRef.get().unlock(lockId);
-            lockRef = lock.lockRef;
-            lockId = lock.lockRef.get().lock();
+            mutex_ref.get().unlock(lock_id);
+            mutex_ref = rhs_lock.mutex_ref;
+            lock_id = rhs_lock.lock_id;
+            rhs_lock.lock_id = 0;    // invalidating rhs_lock
         }
 
         void lock(std::size_t begin, std::size_t end)
         {
-            lockId = lockRef.get().lock(begin, end);
+            lock_id = mutex_ref.get().lock(begin, end);
```

Comment:
What should happen if `lock_id != 0` before the locking operation? Also, should `lock_id` be an atomic?

Reply:
The [cppref](https://en.cppreference.com/w/cpp/thread/unique_lock/lock#exceptions) page for unique_lock mentions the following

> Exceptions
Any exceptions thrown by mutex()->lock().
If there is no associated mutex, [std::system_error](https://en.cppreference.com/w/cpp/error/system_error) with an error code of [std::errc::operation_not_permitted](https://en.cppreference.com/w/cpp/error/errc).
If the mutex is already locked by this unique_lock (in other words, [owns_lock](https://en.cppreference.com/w/cpp/thread/unique_lock/owns_lock) is true), [std::system_error](https://en.cppreference.com/w/cpp/error/system_error) with an error code of [std::errc::resource_deadlock_would_occur](https://en.cppreference.com/w/cpp/error/errc).


I guess we should do the same

Reply:
You are certainly right, I misunderstood things.